libraries/AP_UAVCAN/AP_UAVCAN.cpp

```diff
@@ -62,9 +64,227 @@ const AP_Param::GroupInfo AP_UAVCAN::var_info[] = {
     // @User: Advanced
     AP_GROUPINFO("ESC_BM", 3, AP_UAVCAN, _esc_bm, 255),
 
+    // @Param: BRCST_BM
+    // @DisplayName: Bitmask to set broadcasted messages
+    // @Description: Bitmask with one set message to be broadcasted over UAVCAN. Reboot needed after change.
+    // @Bitmask: 0: Position Fix 1, 1: Position Fix2 2, 1: Attitude 4
+    // @User: Advanced
+    AP_GROUPINFO("BR_BM", 4, AP_UAVCAN, _broadcast_bm, 0),
+
+    // @Param: NODE
+    // @DisplayName: UAVCAN GNSS Fix broadcast rate
+    // @Description: UAVCAN GNSS Fix broadcast rate in times per second. It is active only if enabled with bitmask.
+    // @Range: 1 20
+    // @User: Advanced
+    AP_GROUPINFO("BR_FIX_R", 5, AP_UAVCAN, _broadcast_fix_rate, 10),
+
+    // @Param: NODE
+    // @DisplayName: UAVCAN GNSS Fix2 broadcast rate
+    // @Description: UAVCAN GNSS Fix2 broadcast rate in times per second. It is active only if enabled with bitmask.
+    // @Range: 1 20
+    // @User: Advanced
+    AP_GROUPINFO("BR_FIX2_R", 6, AP_UAVCAN, _broadcast_fix2_rate, 10),
+
+    // @Param: NODE
+    // @DisplayName: UAVCAN attitude broadcast rate
+    // @Description: UAVCAN attitude broadcast rate in times per second. It is active only if enabled with bitmask.
+    // @Range: 1 20
+    // @User: Advanced
+    AP_GROUPINFO("BR_ATT_R", 7, AP_UAVCAN, _broadcast_att_rate, 10),
+
     AP_GROUPEND
 };
 
+// publisher interfaces
+static uavcan::Publisher<uavcan::equipment::gnss::Fix> *fix_out_array[MAX_NUMBER_OF_CAN_DRIVERS];
+static uavcan::Publisher<uavcan::equipment::gnss::Fix2> *fix2_out_array[MAX_NUMBER_OF_CAN_DRIVERS];
+static uavcan::Publisher<uavcan::equipment::ahrs::Solution> *attitude_out_array[MAX_NUMBER_OF_CAN_DRIVERS];
+
+static uavcan::equipment::gnss::Fix _fix_state;
+static uavcan::equipment::gnss::Fix2 _fix2_state;
+static uavcan::equipment::ahrs::Solution _att_state;
+
+void AP_UAVCAN::UAVCAN_AHRS_update(const AP_AHRS_NavEKF &ahrs)
+{
+    if (fix_out_array[_uavcan_i] != nullptr) {
+        bool sem_ret = _fix_out_sem->take(1);
```

Comment:
Why use 2 semaphores for this? As far as I can tell you only execute this code if you can grab both of them, and the only other use site I can find grabs one then the other. Because you are trying to protect this one location wouldn't a single semaphore be sufficient?

Reply:
Fixed